### PR TITLE
Add duration.

### PR DESCRIPTION
**Pull request by Shylock-Hg**
#### What type of PR is this?
- [ ] bug
- [x] feature
- [ ] enhancement

#### What does this PR do?
Related to #3101 

#### Checklist：
- [x] Documentation affected （Please add the label if documentation needs to be modified.)
- [x] Incompatible （If it is incompatible, please describe it and add corresponding label.）
- [ ] Need to cherry-pick （If need to cherry-pick to some branches, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes：
Please confirm whether to reflect in release notes and how to describe:
>                                                                 `

### CODE REVIEW

File: src/common/datatypes/Date.cpp [hunk 1]
```diff
@@ -92,16 +97,224 @@ Date Date::operator-(int64_t days) const {
   return Date(daysSince - days);
 }
 
+void Date::addDuration(const Duration& duration) {
```
Comment (czpmango) [inline src/common/datatypes/Date.cpp:100]: Better to modify this function to `friend Date operator+(const Date &date, const Duration &dur)`.
eg. The implementation of expression `Duration(..)+Date(..)` will be simpler.

File: src/common/datatypes/Date.cpp [hunk 2]
```diff
@@ -92,16 +97,224 @@ Date Date::operator-(int64_t days) const {
...
+  } else {
+    hour = tmp;
+  }
+  if (hour < 0) {
+    carry--;
+    hour += 24;
+  }
+}
+
+void Time::subDuration(const Duration& duration) { addDuration(-duration); }
```
Comment (czpmango) [inline src/common/datatypes/Date.cpp:200]: Ditto.

File: src/common/datatypes/Date.cpp (hunk 1 above)
Comment (Shylock-Hg) [reply src/common/datatypes/Date.cpp:100]: Fixed.

**Review by czpmango — DISMISSED** [review]

**Review by czpmango — APPROVED** [review]

**Review by yixinglu — COMMENTED** [review]
What happens if we insert a duration type data?

**Comment by Shylock-Hg** [comment]
> What happens if we insert a duration type data?

Insert error in current.

**Comment by yixinglu** [comment]
> > What happens if we insert a duration type data?
> 
> 
> 
> Insert error in current.

could you add some tests for the scenario?

**Comment by Shylock-Hg** [comment]
> > > What happens if we insert a duration type data?
> > 
> > 
> > Insert error in current.
> 
> could you add some tests for the scenario?

Added. And support to store it.

**Review by czpmango — APPROVED** [review]

**Review by czpmango — APPROVED** [review]

File: tests/tck/features/expression/TimeComputation.feature [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      | <sum> | <diff> |
+
+    Examples:
+      | map                                                                   | sum          | diff         |
+      | {years: 12, months: 5, days: 14, hours: 16, minutes: 12, seconds: 70} | '1997-03-25' | '1972-04-27' |
+
+  # TODO don't allow different sign
+  # | {months: 1, days: -14, hours: 16, minutes: -12, seconds: 70}                                      | '1984-10-28' | '1984-09-25' |
+  # TODO support decimal
+  # | {years: 12.5, months: 5.5, days: 14.5, hours: 16.5, minutes: 12.5, seconds: 70.5} | '1997-10-11' | '1971-10-12' |
```
Comment (CPWstatic) [inline tests/tck/features/expression/TimeComputation.feature:26]: What's the behavior for these cases now?

**Review by yixinglu — APPROVED** [review]

**Pull request merged**